src/librustdoc/lib.rs

```diff
@@ -112,7 +118,45 @@ mod theme;
 mod visit_ast;
 mod visit_lib;
 
+// FIXME(guswynn): copy documentation from rustc/src/main.rs
+#[cfg(feature = "jemalloc")]
+#[global_allocator]
+static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
+
 pub fn main() {
+    #[cfg(feature = "jemalloc")]
+    {
+        use std::os::raw::{c_int, c_void};
+
+        #[used]
+        static _F1: unsafe extern "C" fn(usize, usize) -> *mut c_void = jemalloc_sys::calloc;
+        #[used]
+        static _F2: unsafe extern "C" fn(*mut *mut c_void, usize, usize) -> c_int =
+            jemalloc_sys::posix_memalign;
+        #[used]
+        static _F3: unsafe extern "C" fn(usize, usize) -> *mut c_void = jemalloc_sys::aligned_alloc;
+        #[used]
+        static _F4: unsafe extern "C" fn(usize) -> *mut c_void = jemalloc_sys::malloc;
+        #[used]
+        static _F5: unsafe extern "C" fn(*mut c_void, usize) -> *mut c_void = jemalloc_sys::realloc;
+        #[used]
+        static _F6: unsafe extern "C" fn(*mut c_void) = jemalloc_sys::free;
+
+        // On OSX, jemalloc doesn't directly override malloc/free, but instead
+        // registers itself with the allocator's zone APIs in a ctor. However,
+        // the linker doesn't seem to consider ctors as "used" when statically
+        // linking, so we need to explicitly depend on the function.
+        #[cfg(target_os = "macos")]
+        {
+            extern "C" {
+                fn _rjem_je_zone_register();
+            }
+
+            #[used]
+            static _F7: unsafe extern "C" fn() = _rjem_je_zone_register;
+        }
+    }
+
```

Comment:
not sure why this was originally in `rustc` and not `rustc_driver`, maybe it only works if its in the leaf binary crate? @alexcrichton is the origin of this I believe?

Reply:
The movement happened in https://github.com/rust-lang/rust/pull/56986 in response to https://github.com/rust-lang/rust/issues/56980. I believe it boils down to, at the time, "the RLS didn't do well with jemalloc". I don't believe the reason for "not doing well" was ever investigated, however.

Reply:
That encouraging to hear that it should correctly overwrite the symbols (and we can move to `src/tools/rustdoc/src/main.rs` if we need to for RLS (though I believe most people use rust-analyzer now)

Reply:
> I don't believe the reason for "not doing well" was ever investigated, however.

I don't think it was and IIRC the main problem was memory characteristics when running multiple `rustc` in-process in the RLS, which is not done anymore by default (we spawn child processes instead).

Feel free to disregard RLS when making changes here.